Universal property of smash products (#865)
UniMath/agda-unimath: src/synthetic-homotopy-theory/smash-products-of-pointed-types.lagda.md

> ### The canonical pointed map from the product to the smash product

Reviewer: ```suggestion
### The canonical pointed map from the cartesian product to the smash product
```
Seems reasonable to be specific about which product we mean precisely.

    pr1 (gap-smash-prod-Pointed-Type f g)
```

### The canonical map from the wedge sum to the smash product identifies all points

```agda
module _
  {l1 l2 : Level} (A : Pointed-Type l1) (B : Pointed-Type l2)
  where

  pointed-map-smash-prod-wedge-Pointed-Type : (A ∨∗ B) →∗ (A ∧∗ B)
  pointed-map-smash-prod-wedge-Pointed-Type =
    pointed-map-smash-prod-prod-Pointed-Type A B ∘∗
    pointed-map-prod-wedge-Pointed-Type A B

  map-smash-prod-wedge-Pointed-Type :
    type-Pointed-Type (A ∨∗ B) → type-Pointed-Type (A ∧∗ B)
  map-smash-prod-wedge-Pointed-Type =
    map-pointed-map pointed-map-smash-prod-wedge-Pointed-Type

>   eq-map-smash-prod-wedge-Pointed-Type :
>     ( x : type-Pointed-Type (A ∨∗ B)) →
>     map-smash-prod-wedge-Pointed-Type x ＝
>     point-Pointed-Type (A ∧∗ B)
>   eq-map-smash-prod-wedge-Pointed-Type x =
>     ( glue-pushout
>       ( map-prod-wedge-Pointed-Type A B)
>       ( map-pointed-map {A = A ∨∗ B} {B = unit-Pointed-Type}
>         ( terminal-pointed-map (A ∨∗ B)))
>       ( x)) ∙
>     ( ( htpy-right-whisk
>         ( htpy-pointed-htpy
>           ( inr-pushout-Pointed-Type
>             ( pointed-map-prod-wedge-Pointed-Type A B)
>             ( terminal-pointed-map (A ∨∗ B)))
>           ( inclusion-point-Pointed-Type (A ∧∗ B))
>           ( is-initial-unit-Pointed-Type
>             ( A ∧∗ B)
>             ( inr-pushout-Pointed-Type
>               ( pointed-map-prod-wedge-Pointed-Type A B)
>               ( terminal-pointed-map (A ∨∗ B)))))
>         ( map-terminal-pointed-map (A ∨∗ B)))
>       ( x)) ∙
>     ( preserves-point-pointed-map
>       ( inclusion-point-Pointed-Type (A ∧∗ B)))

Reviewer: Removes a parenthesis and puts some things on the same line
```suggestion
  eq-map-smash-prod-wedge-Pointed-Type :
    (x : type-Pointed-Type (A ∨∗ B)) →
    map-smash-prod-wedge-Pointed-Type x ＝ point-Pointed-Type (A ∧∗ B)
  eq-map-smash-prod-wedge-Pointed-Type x =
    ( glue-pushout
      ( map-prod-wedge-Pointed-Type A B)
      ( map-pointed-map {A = A ∨∗ B} {B = unit-Pointed-Type}
        ( terminal-pointed-map (A ∨∗ B)))
      ( x)) ∙
    ( htpy-right-whisk
      ( htpy-pointed-htpy
        ( inr-pushout-Pointed-Type
          ( pointed-map-prod-wedge-Pointed-Type A B)
          ( terminal-pointed-map (A ∨∗ B)))
        ( inclusion-point-Pointed-Type (A ∧∗ B))
        ( is-initial-unit-Pointed-Type
          ( A ∧∗ B)
          ( inr-pushout-Pointed-Type
            ( pointed-map-prod-wedge-Pointed-Type A B)
            ( terminal-pointed-map (A ∨∗ B)))))
        ( map-terminal-pointed-map (A ∨∗ B))
      ( x)) ∙
    ( preserves-point-pointed-map (inclusion-point-Pointed-Type (A ∧∗ B)))
```

Reviewer 2: It also unindents the `x : ` in the arguments; we don't have a good (or any whatsoever) guideline for this formatting, but I'm always writing a space between the opening paranthesis and the name of the argument for symmetry with function application.

Reviewer: True. I usually prefer to not have that space, but we should really get around to formulating a more precise convention.

Reviewer: Jonathan would agree with your suggestion btw, and if our convention were to be to always have spaces after parentheses on new lines, then that would be easy to implement an autoformatter for (that he wrote).

Reviewer: Egbert, Jonathan, and I had a discussion about this in the past, and landed on that we disagree on whether this should be the rule.

Reviewer 2: Ah, thanks for the historical background, I wasn't sure if this was just overlooked or if there had already been a discussion on that topic 👍

Author: Right, I had in my mind while I was writing this that a homotopy and its argument need to lie on the same level and with that thinking it seemed wrong to have the `( x)` on the same line as the other steps in the proof. Thanks for the fix!

Author: Otherwise I think I agree with Vojtech that having the space in the arguments is nice. I've gotten so used to putting spaces before function applications that something looks wrong without them everywhere else.

Reviewer: I went looking for the initial discussion, and it turns out we had it in a private chat on Discord, so sorry, but I don't have anything to reference back to. I'll wait a little and see if Egbert wants to weigh in, but I am still of the viewpoint that the added space adds nothing, and I just personally find it looks awkward. Although maybe if it was enforced consistently that feeling would change.
Also, I just realized, it shouldn't be hard to write an autoformatting hack regardless if the convention is to have a space or not. I guess if we really wanted, we could try both shoes on to see what fits best.

Reviewer: Anyway, I edited the suggestion to omit the controversial change

      ( glue-wedge-Pointed-Type A B)) ∙
    ( eq-map-smash-prod-wedge-Pointed-Type
        ( map-inr-wedge-Pointed-Type A B (point-Pointed-Type B))) ＝
    ( eq-map-smash-prod-wedge-Pointed-Type
      ( map-inl-wedge-Pointed-Type A B (point-Pointed-Type A)))
  coh-eq-map-smash-prod-wedge-Pointed-Type =
    ( inv-map-compute-dependent-identification-eq-value-function
      ( map-smash-prod-wedge-Pointed-Type)
      ( map-pointed-map
        ( constant-pointed-map (A ∨∗ B) (A ∧∗ B)))
      ( glue-wedge-Pointed-Type A B)
      ( eq-map-smash-prod-wedge-Pointed-Type
        ( map-inl-wedge-Pointed-Type A B (point-Pointed-Type A)))
      ( eq-map-smash-prod-wedge-Pointed-Type
        ( map-inr-wedge-Pointed-Type A B (point-Pointed-Type B)))
      ( apd
        ( eq-map-smash-prod-wedge-Pointed-Type)
        ( glue-wedge-Pointed-Type A B))) ∙
    ( identification-left-whisk
      ( eq-map-smash-prod-wedge-Pointed-Type
        ( map-inl-wedge-Pointed-Type A B (point-Pointed-Type A)))
      ( ap-const
        ( point-Pointed-Type (A ∧∗ B))
        ( glue-wedge-Pointed-Type A B))) ∙
    ( right-unit)
```

### The map from the pointed product to the smash product identifies elements

> of the form (x, b) and (a, y).

Reviewer: ```suggestion
of the form `(x, b)` and `(a, y)`, where `b` and `a` are the base points of `B`
and `A` respectively.
```

```agda
module _
  {l1 l2 : Level}
  (A : Pointed-Type l1) (B : Pointed-Type l2)
  where

  glue-smash-prod-Pointed-Type :
    (x : type-Pointed-Type A) (y : type-Pointed-Type B) →
    map-smash-prod-prod-Pointed-Type A B
      ( x , point-Pointed-Type B) ＝
    map-smash-prod-prod-Pointed-Type A B
      ( point-Pointed-Type A , y)
  glue-smash-prod-Pointed-Type x y =
    ( ap
      ( map-smash-prod-prod-Pointed-Type A B)
      ( inv (inl-prod-wedge-Pointed-Type A B x))) ∙
    ( eq-map-smash-prod-wedge-Pointed-Type A B
      ( map-inl-wedge-Pointed-Type A B x) ∙
    ( inv
      ( eq-map-smash-prod-wedge-Pointed-Type A B
        ( map-inr-wedge-Pointed-Type A B y))) ∙
    ( ap
      ( map-smash-prod-prod-Pointed-Type A B)

>       ( inr-prod-wedge-Pointed-Type A B y)))

Reviewer: This is either misindented or misparenthesized.

          ( inr-prod-wedge-Pointed-Type A B (point-Pointed-Type B))))) ∙
    ( inv
      ( ap-concat
        ( map-smash-prod-prod-Pointed-Type A B)
        ( inv (inl-prod-wedge-Pointed-Type A B (point-Pointed-Type A)))
        ( ( ap
            ( map-prod-wedge-Pointed-Type A B)
            ( glue-wedge-Pointed-Type A B)) ∙
          ( inr-prod-wedge-Pointed-Type A B (point-Pointed-Type B))))) ∙
    ( ap²
      ( map-smash-prod-prod-Pointed-Type A B)
      ( inv
        ( left-transpose-eq-concat
          ( inl-prod-wedge-Pointed-Type A B (point-Pointed-Type A))
          ( refl)
          ( ( ap
              ( map-prod-wedge-Pointed-Type A B)
              ( glue-wedge-Pointed-Type A B) ∙
            ( inr-prod-wedge-Pointed-Type A B (point-Pointed-Type B))))
          ( inv
            ( compute-glue-cogap
              ( map-pointed-map
                ( inclusion-point-Pointed-Type A))
              ( map-pointed-map
                ( inclusion-point-Pointed-Type B))
              ( cocone-type-cocone-Pointed-Type
                ( inclusion-point-Pointed-Type A)
                ( inclusion-point-Pointed-Type B)
                ( cocone-prod-wedge-Pointed-Type A B))

>               ( point-Pointed-Type unit-Pointed-Type))))))

Reviewer: There is some fishy parenthisation going on in this proof. Here's a rule that should be followed: When a path concatenation is part of a multiline expression, each argument should be parenthesized, so, for instance, in the line that reads `( p ∙` (end of line), the `p` should be parenthesized.

Moreover when the association of the proof isn't relevant, I/we prefer to omit the parentheses. So for instance if you are proving `x = y` by `p ∙ (q ∙ r)`, and you don't care about how that proof is constructed later, I would omit the parentheses.

Reviewer: There's also some indentation issues, but I'm sure you will spot those when you fix the first problem. You can spot it by how you have written

```agda
  ( p ∙
  ( ...
```

What helps me spot such issues quickly is VSCode's rainbow parentheses setting. It will display the two parentheses with different colors, so I know they should be at different indentation levels. I can recommend looking for a similar feature in your preferred editor!

Reviewer: Rainbow parentheses also help with all sorts of different issues that arise frequently when writing agda-unimath code anyway, like spotting overparenthization.

Reviewer: Here, it seems like someone may end up caring about how this particular coherence proof is constructed, so you can disregard my comment about association.

Author: > hen again, since this coherence is data and not just a property, you might end up caring about the particular construction that is made. So then it is worthwhile to consider if there is a more direct or streamlined proof that can be made.

Thanks for these comments! The association was in fact on purpose, because I couldn't get something to work with the usual formatting. Thinking about this proof more made me realize there were potentially some issues in the way it was constructed though, and I've pushed a fix now. I think everything should also follow the usual parenthesization now :)

Reviewer: Excellent, that's quite a bit better! Good work :)

```

### The universal property of the smash product

> ```agda

Reviewer: It would be nice with a short explanation of what the universal property is. Would you mind writing that? It should also be mentioned that it is not the categorical product and that the cartesian product is.

Reviewer: Something simple like the following should suffice:
````suggestion
### The universal property of the smash product

The _universal property of the smash product_ is that it forms the left-adjoint
to [pointed functions](structured-types.pointed-maps.md) in pointed types:

```text
  ((A ∧∗ B) →∗ C) ≃∗ (A →∗ B →∗ C).
```

**Note:** The categorical product in pointed types is the
[pointed cartesian product](structured-types.pointed-cartesian-product-types.md).

```agda
````

Please do confirm that I'm not spewing nonsense again, though.

Author: That's right! I decided to expand it a little bit though :)

map-universal-property-smash-prod-Pointed-Type :
  {l1 l2 l3 : Level}
  (A : Pointed-Type l1) (B : Pointed-Type l2) (C : Pointed-Type l3) →
  ((A ∧∗ B) →∗ C) → (type-Pointed-Type A) → (B →∗ C)
pr1 (map-universal-property-smash-prod-Pointed-Type A B C f x) y =
  map-pointed-map f (map-smash-prod-prod-Pointed-Type A B (x , y))
pr2 (map-universal-property-smash-prod-Pointed-Type A B C f x) =
  ( ap
    ( map-pointed-map f)
    ( glue-smash-prod-Pointed-Type A B x (point-Pointed-Type B))) ∙
  ( preserves-point-pointed-map f)

universal-property-smash-prod-Pointed-Type :
  {l1 l2 l3 : Level}
  (A : Pointed-Type l1) (B : Pointed-Type l2) (C : Pointed-Type l3) →
  ((A ∧∗ B) →∗ C) → (A →∗ (pointed-map-Pointed-Type B C))
pr1 (universal-property-smash-prod-Pointed-Type A B C f) =
  map-universal-property-smash-prod-Pointed-Type A B C f
pr2 (universal-property-smash-prod-Pointed-Type A B C f) =
  eq-htpy-pointed-Π
    ( map-universal-property-smash-prod-Pointed-Type A B C
      ( f)
      ( point-Pointed-Type A))
    ( constant-pointed-map B C)
    ( ( λ y →
        ( ap
          ( map-pointed-map f)

>           ( inv
>             ( glue-smash-prod-Pointed-Type A B (point-Pointed-Type A) y)) ∙

Reviewer: ```suggestion
          ( inv (glue-smash-prod-Pointed-Type A B (point-Pointed-Type A) y)) ∙
```

    pr1 (gap-smash-prod-Pointed-Type f g)
```

### The canonical map from the wedge sum to the smash product identifies all points

```agda
module _
  {l1 l2 : Level} (A : Pointed-Type l1) (B : Pointed-Type l2)
  where

  pointed-map-smash-prod-wedge-Pointed-Type : (A ∨∗ B) →∗ (A ∧∗ B)
  pointed-map-smash-prod-wedge-Pointed-Type =
    pointed-map-smash-prod-prod-Pointed-Type A B ∘∗
    pointed-map-prod-wedge-Pointed-Type A B

  map-smash-prod-wedge-Pointed-Type :
    type-Pointed-Type (A ∨∗ B) → type-Pointed-Type (A ∧∗ B)
  map-smash-prod-wedge-Pointed-Type =
    map-pointed-map pointed-map-smash-prod-wedge-Pointed-Type

  eq-map-smash-prod-wedge-Pointed-Type :
    ( x : type-Pointed-Type (A ∨∗ B)) →
    map-smash-prod-wedge-Pointed-Type x ＝
    point-Pointed-Type (A ∧∗ B)

>   eq-map-smash-prod-wedge-Pointed-Type x =

Reviewer: Yeah, this name should also change, but it doesn't read like a `compute` to me, more like a contraction. It is precisely showing that the image of the map contracts at the base point. But doesn't that also mean that the accompanying coherence (and all higher coherences) are automatic? Anyway, perhaps a good name is `contraction-map-smash-prod-...`?

Reviewer: We already know the map is pointed. So specifying where the contraction is based at in the name seems redundant.

Reviewer: Right, you can disregard that. I'm probably just seeing things 🙃

Author: In classical algebraic topology I've heard functions that are homotopic to a constant map called "null homotopic". This terminology isn't in use in the library, right? Peraphs 'is-null-homotopic-map-smash-prod-...` would be a good name?

Author: I guess that would pretty strongly imply that what this function computes is something like `is-null-homotopic (map-smash-prod...)`. I think the name you suggested should be ok.

Reviewer: Your suggestion is also good!
